### DISCUSSION TITLE
handle ping/pong in ws

### DISCUSSION THREAD

**Discussion by hosseinhojjatyar** [uNetworking/uWebSockets.js#394] [Q&A] [answered]
hi,
how handle ping/pong in ws?
i can't find an example in [https://github.com/uNetworking/uWebSockets.js/tree/master/examples](https://github.com/uNetworking/uWebSockets.js/tree/master/examples)

**Comment by hst-m** [comment, marked as answer]
v19 automatically sends Ping to client before the idleTimeout to keep connection alive, or you can manually send Ping with ws.ping()

Sending Ping from server: [websocket.html#ping](https://unetworking.github.io/uWebSockets.js/generated/interfaces/websocket.html#ping)
Receiving Ping event: [websocketbehavior.html#ping](https://unetworking.github.io/uWebSockets.js/generated/interfaces/websocketbehavior.html#ping)
Receiving Pong event: [websocketbehavior.html#pong](https://unetworking.github.io/uWebSockets.js/generated/interfaces/websocketbehavior.html#pong)

**Reply by buildscientist** [reply]
@hst-m 
Does the websocket.ping() method override default behavior for server side pings in v19.x? I want to be able to add some timing metrics for latency measurement between my UWS server and the client responding to ping event.

**Reply by e3dio** [reply]
Calling ws.ping() on the server will make the client send Pong back, Pong handler will trigger on server so you can add your latency measurement there, receiving the pong will reset the idleTimeout and prevent server from auto-sending ping to client. You can also set ws option `sendPingsAutomatically` to false if you don't want it auto-sending pings at idleTimeout. Docs need to be updated with the extra options

**Reply by lewispham** [reply]
@e3dio  As mentioned in the docs, why do we need to specify a number instead of a boolean value for `sendPingsAutomatically`? When that option is enabled, how and when will the pings be sent?

**Reply by e3dio** [reply]
Docs need to be fixed, see https://github.com/uNetworking/uWebSockets.js/issues/694

**Comment by dragonbane0** [comment]
@hst-m Thank you a lot for all these little snippets btw. Helps a lot starting out :)
